ssl/t1_lib.c

```diff
@@ -417,6 +417,314 @@ int ssl_load_groups(SSL_CTX *ctx)
+        if (sinf->hash_name == NULL)
+            goto err;
+    }
+
+    p = OSSL_PARAM_locate_const(params, OSSL_CAPABILITY_TLS_SIGALG_HASH_OID);
+    if (p == NULL)
+        sinf->hash_oid = NULL;
+    else if (p->data_type != OSSL_PARAM_UTF8_STRING)
+        goto err;
+    else {
+        sinf->hash_oid = OPENSSL_strdup(p->data);
+        if (sinf->hash_oid == NULL)
+            goto err;
+    }
+
+    p = OSSL_PARAM_locate_const(params, OSSL_CAPABILITY_TLS_SIGALG_KEYTYPE);
+    if (p == NULL)
+        sinf->keytype = NULL;
+    else if (p->data_type != OSSL_PARAM_UTF8_STRING)
+        goto err;
+    else {
+        sinf->keytype = OPENSSL_strdup(p->data);
+        if (sinf->keytype == NULL)
+            goto err;
+    }
+
+    p = OSSL_PARAM_locate_const(params, OSSL_CAPABILITY_TLS_SIGALG_KEYTYPE_OID);
+    if (p == NULL)
+        sinf->keytype_oid = NULL;
+    else if (p->data_type != OSSL_PARAM_UTF8_STRING)
+        goto err;
+    else {
+        sinf->keytype_oid = OPENSSL_strdup(p->data);
+        if (sinf->keytype_oid == NULL)
+            goto err;
+    }
+
+    /* The remaining parameters below are mandatory again */
+    p = OSSL_PARAM_locate_const(params, OSSL_CAPABILITY_TLS_SIGALG_MIN_TLS);
+    if (p == NULL || !OSSL_PARAM_get_int(p, &sinf->mintls)) {
+        ERR_raise(ERR_LIB_SSL, ERR_R_PASSED_INVALID_ARGUMENT);
+        goto err;
+    }
+    if ((sinf->mintls != 0) && (sinf->mintls != -1) &&
+        ((sinf->mintls < TLS1_VERSION))) {
+        ERR_raise(ERR_LIB_SSL, ERR_R_PASSED_INVALID_ARGUMENT);
+        goto err;
+    }
+
+    p = OSSL_PARAM_locate_const(params, OSSL_CAPABILITY_TLS_SIGALG_MAX_TLS);
+    if (p == NULL || !OSSL_PARAM_get_int(p, &sinf->maxtls)) {
+        ERR_raise(ERR_LIB_SSL, ERR_R_PASSED_INVALID_ARGUMENT);
+        goto err;
+    }
+    if ((sinf->maxtls != 0) && (sinf->maxtls != -1) &&
+        ((sinf->maxtls < sinf->mintls)
+         || (sinf->maxtls > TLS_MAX_VERSION_INTERNAL))) {
+        ERR_raise(ERR_LIB_SSL, ERR_R_PASSED_INVALID_ARGUMENT);
+        goto err;
+    }
+
+    p = OSSL_PARAM_locate_const(params, OSSL_CAPABILITY_TLS_SIGALG_MIN_DTLS);
+    if (p == NULL || !OSSL_PARAM_get_int(p, &sinf->mindtls)) {
+        ERR_raise(ERR_LIB_SSL, ERR_R_PASSED_INVALID_ARGUMENT);
+        goto err;
+    }
+
+    p = OSSL_PARAM_locate_const(params, OSSL_CAPABILITY_TLS_SIGALG_MAX_DTLS);
+    if (p == NULL || !OSSL_PARAM_get_int(p, &sinf->maxdtls)) {
+        ERR_raise(ERR_LIB_SSL, ERR_R_PASSED_INVALID_ARGUMENT);
+        goto err;
+    }
+    /*
+     * Now check that the algorithm is actually usable for our property query
+     * string. Regardless of the result we still return success because we have
+     * successfully processed this signature, even though we may decide not to
+     * use it.
+     */
+    ret = 1;
+    ERR_set_mark();
+    keymgmt = EVP_KEYMGMT_fetch(ctx->libctx, sinf->keytype?sinf->keytype:
+		    sinf->sig_name?sinf->sig_name:sinf->sigalg_name, ctx->propq);
```

Comment:
```suggestion
    ERR_set_mark();
    keytype = sinf->keytype ? sinf->keytype : sinf->sig_name ? sinf->sig_name : sinf->sigalg_name;
    keymgmt = EVP_KEYMGMT_fetch(ctx->libctx, keytype, ctx->propq);
```

Reply:
Hmm, did I convince or confuse you to not use the `(sinf->sig_name != NULL)` style? Would be glad to commit your suggestion as-is :)

Reply:
I got confused.  our coding style is quite clear, look for yourself:

https://github.com/openssl/technical-policies/blob/4fc15576a57ee6387dfe58ee241bda3e40e2f06c/policies/coding-style.md?plain=1#L684-L703

So just give me a second to edit the review comment, yeah?

Reply:
There

Reply:
Thanks!